[PC-1601] Modbus TCP | PMC + Opta | Written tutorial (#1700)
arduino/docs-content: content/hardware/05.pro-solutions/solutions-and-kits/portenta-machine-control/tutorials/pmc-opta-modbus-tcp/content.md

Reviewer: Thanks, yeah, I will consider this for next media.

---
title: 'Modbus TCP with Portenta Machine Control & Opta™'
difficulty: intermediate

> description: "Modbus TCP communication on a real industrial application using a Portenta Machine Control, Opta™, a temperature sensor, and the Arduino® PLC IDE."

Reviewer 2: I miss the action

> description: "Modbus TCP communication on a real industrial application using a Portenta Machine Control, Opta™, a temperature sensor, and the Arduino® PLC IDE."

Reviewer 2: What about

```suggestion
description: " Create a Modbus TCP communication on a real industrial application using a Portenta Machine Control, Opta™, a temperature sensor, and the Arduino® PLC IDE."
```

> description: "Modbus TCP communication on a real industrial application using a Portenta Machine Control, Opta™, a temperature sensor, and the Arduino® PLC IDE."

Reviewer: I will use instead "Learn to use..."

#### Modbus TCP - Server
For the Modbus TCP configuration, on the **resources tab** go to the **Ethernet** section. As noticed, the Modbus TCP Slave mode is always enabled, so you don't have to make any changes.

![Modbus set to slave mode](assets/slave-mode.png)

Now, go to the **sketch editor** and uncomment the library and setup function code lines. As the IP, we must use the same as the Portenta Machine Control, as it is connected to your router, you can find it on its configurations. 

In this case, the following configurations are used.

```arduino
// Enable usage of EtherClass, to set static IP address and other
#include <PortentaEthernet.h>
arduino::EthernetClass eth(&m_netInterface);

void setup()
{
	// Configure static IP address
	IPAddress ip(10, 0, 0, 157);    // PMC IP address
	IPAddress dns(10, 0, 0, 1);     // gateway IP address
	IPAddress gateway(10, 0, 0, 1); // gateway IP address
	IPAddress subnet(255, 255, 255, 0); 
	// If cable is not connected this will block the start of PLC with about 60s of timeout!
	eth.begin(ip, dns, gateway, subnet);

}
```
![Network settings for Modbus TCP](assets/ip-setup.png)

> Now, create the variable that will be shared with the temperature sensor data in the network. For this, we go to **status variables** and click on **Add**. Configure it as follows:

Reviewer 2: ```suggestion
Now create the variable that will be shared with the temperature sensor data in the network. Go to **status variables** and click on **Add**. Configure it as follows:
```

- Name: `temp_send`
- Address: `25000`
- PLC type: `REAL`

![Temperature variable setup](assets/var-setup.png)

> Next, go to **Temperature probes** and select the sensor type, for this tutorial, enable the **thermocouple** connected to the first channel by setting it to true.

Reviewer 2: ```suggestion
Next go to **Temperature probes** and select the sensor type, for this tutorial, enable the **thermocouple** connected to the first channel by setting it to true.
```

![Temperature probe configuration](assets/probe-setup.png)

Finally, go to the **main program** in **project** and match the temperature variable with the temperature sensor lectures as follows:

```
temp_send := sysTempProbes[0].temperature; 
```

![Main program for the server](assets/main-code.png)

To check if everything is okay, click on compile, and if no error is shown, you can upload the code to your Portenta Machine Control. 

![Project compiled with no errors](assets/compile.png)

> Once uploaded, click on the **Connect** button again. Now, you can monitor the **temp_send** variable in the **Watch** window dragging and dropping it from the __Global shared__ variables in the project tab.

Reviewer 2: ```suggestion
Once uploaded, click on the **Connect** button again. Now you can monitor the **temp_send** variable in the **Watch** window dragging and dropping it from the __Global shared__ variables in the project tab.
```

> You should see the temperature value measured by the sensor.

Reviewer 2: ```suggestion
You should see the temperature value measured by the sensor inside the "Watch" window.
```

![Live temperature measurement on the PMC](assets/temp-check.gif)

### Opta Micro PLC Setup

Now the server is configured, create a new project, this time for the Opta™ micro PLC that will be the Client or Master.

![New project for the Opta™](assets/new-project-opta.png)

Upload the runtime for Opta™ by selecting its serial port and clicking on the **Download** button as before.

![Uploading the runtime to the Opta™](assets/runtime-opta.png)

Once the runtime is flashed, with your Opta™ connected to your router, search for its IP address on the router configurations.

On the PLC IDE, navigate to **On-line > Set up communication**, activate and then open the **ModbusTCP** properties. Add the Opta™ IP address, then click "OK". 

![Modbus TCP connection](assets/modbus-prog-opta.png)
![Modbus TCP IP setup](assets/modbus-ip-opta.png)

Now, in the upper left corner, click on the **Connect** button and wait for the base program to be uploaded. A green **Connected** flag should appear in the lower right corner if everything goes well.

![Connecting the board](assets/connected-opta.png)

> ***The Opta™ doesn’t need any license activation to be used with the PLC IDE***

Reviewer 2: Are we sure about this? there are still some steps to do with pre-licensed products like Opta:

https://docs.arduino.cc/tutorials/portenta-machine-control/plc-ide-setup-license/?queryID=undefined#7-license-activation-with-pre-licensed-products-opta

Please check and update the tutorial accordingly

Reviewer: Noted! Thanks

#### Modbus TCP - Client

With the Opta™ successfully connected to the PLC IDE, it is time to configure the Modbus TCP communication.

In the **Ethernet** tab, enable the Modbus TCP Master mode.

![Modbus TCP Master mode enabled](assets/master-modbus.png)

Then right-click on the **Ethernet** tab, click on **Add** and select the _Generic Modbus device_.

> ![Modbus Device Configuration 1](assets/modbus-device.png)

Reviewer 2: ```suggestion
![Modbus Device Configuration - Step 1](assets/modbus-device.png)
```

Not a fan of enumerating images descriptions, but in this case adding the "Step <number>" is much better, please do for the rest of these images of Modbus Device Configuration

On the `Generic Modbus_1` device settings, enter the Server IP, the one from the Portenta Machine Control.

![Modbus Device Configuration 2](assets/modbus-device-2.png)

Right-click on the device and add the **FC-04 Modbus** function that will let us read the server input registers. 

![Modbus Device Configuration 3](assets/modbus-device-3.png)

> Now, click on the function and in the general settings, enter the shared variable address that you defined earlier in the server, `25000` in this case.

Reviewer 2: ```suggestion
Now click on the function and in the general settings, enter the shared variable address that you defined earlier in the server, `25000` in this case.
```

![Modbus Device Configuration 4](assets/modbus-device-4.png)

In the Input Register tab, create a label for it, which could be `temp_reg`.

![Label the register](assets/modbus-device-5.png)

> Now, go to the **sketch editor** and uncomment the library and setup function code lines. As the IP, we must use the same as the Opta™. 

Reviewer 2: ```suggestion
Go to the **sketch editor** and uncomment the library and setup function code lines. As the IP, we must use the same as the Opta™. 
```

> In this case, the following configurations are used.

Reviewer 2: ```suggestion
In this case the following configurations are used:
```

```arduino
// Enable usage of EtherClass, to set static IP address and other
#include <PortentaEthernet.h>
arduino::EthernetClass eth(&m_netInterface);

void setup()
{
	// Configure static IP address
	IPAddress ip(10, 0, 0, 227);    // Opta IP address
	IPAddress dns(10, 0, 0, 1);     // gateway IP address
	IPAddress gateway(10, 0, 0, 1); // gateway IP address
	IPAddress subnet(255, 255, 255, 0); 
	// If cable is not connected this will block the start of PLC with about 60s of timeout!
	eth.begin(ip, dns, gateway, subnet);

}
```
![Network settings for Modbus TCP](assets/ip-setup-opta.png)

> Finally, define the Opta™ outputs behavior in function of the temperature read from the Portenta Machine Control. For this, go to the **resources tab > Relay Outputs** and give a variable name to each relay, in this case, call them `relay_1`, 2, 3 and 4 respectively.

Reviewer 2: ```suggestion
Finally, define the Opta™ outputs behavior in function of the temperature read from the Portenta Machine Control. To do this, go to the **resources tab > Relay Outputs** and give a variable name to each relay, in this case, call them `relay_1`, 2, 3 and 4 respectively.
```

![Outputs definitions](assets/outputs-opta.png)

> The same with the LED outputs, `LED1`, 2, 3 and 4.

Reviewer 2: ```suggestion
The same with the LED outputs, `LED1`, LED2, LED3 and LED4.
```
Put the full names for this case so it is clearer from the technical point of view, I guess, LED2, LED3, LED4

![LEDs definitions](assets/led-opta.png)

> Now, go to the main code in the Project tab. Create a variable by right-clicking on the local variables window and then **insert**, call it **temp_local** and _integer_ as the type.

Reviewer 2: ```suggestion
Now go to the main code in the Project tab. Create a variable by right-clicking on the local variables window and then **insert**, call it **temp_local** and _integer_ as the type.
```

![Create a local variable](assets/variable-opta.png)

In the code editor, match the local variable with the shared one sent by the Portenta Machine Control.

Once stored locally, design the logic to control the outputs as you want. In this case, four different temperature levels will be set to control each output with the temperature rise respectively. Copy and paste the following script into the **main code** section.

> ```

Reviewer 2: FYI this is a technical limitation of the new docs, there is not syntax highlighting for this language.

temp_local:= temp_reg;

IF temp_local >= 30 THEN
    relay_1 := TRUE;
    LED1 := TRUE;    
ELSE
    relay_1 := FALSE;
    LED1 := FALSE;
END_IF;

IF temp_local >= 50 THEN 
    relay_2 := TRUE;
    LED2 := TRUE;
ELSE
    relay_2 := FALSE;
    LED2 := FALSE;
END_IF;

IF temp_local >= 70 THEN 
    relay_3 := TRUE;
    LED3 := TRUE;
ELSE
    relay_3 := FALSE;
    LED3 := FALSE;    
END_IF;

IF temp_local >= 100 THEN 
    relay_4 := TRUE;
    LED4 := TRUE;
ELSE
    relay_4 := FALSE;
    LED4 := FALSE;    
END_IF;

```
![Final code to control the outputs](assets/main-code-opta.png)

With the final code written, compile the project and upload it the the Opta™ micro PLC.

### Final Test

You can leave each device connected separately to the internet router or connect them together directly with one ethernet cable. The first option will let you update the preferred device remotely as you can access it through the local network.

> Now, you can expose the temperature sensor to some heat and monitor it from the PLC IDE. The Opta™ relay outputs and LEDs will close and turn on when the temperature surpasses the programmed thresholds respectively.

Reviewer 2: ```suggestion
Now you can expose the temperature sensor to some heat and monitor it from the PLC IDE. The Opta™ relay outputs and LEDs will close and turn on when the temperature surpasses the programmed thresholds respectively.
```

![Testing the project](assets/final.gif)

#### Conclusion 

> In this tutorial you learned how to communicate two Arduino PRO products using the Modbus TCP protocol, demonstrating a simple application of sharing temperature data to control a device's outputs.

Reviewer 2: ```suggestion
In this tutorial you learned how to communicate two Arduino PRO products using the Modbus TCP protocol, demonstrating a simple application of sharing temperature data to control the outputs of the devices.
```

A small suggestion.

> As you can notice, the configuration process is very straightforward and the results were as expected. 

Reviewer 2: sentence is good, but we cannot end this way a tutorial.

```suggestion
As you can notice, the configuration process is very straightforward and the results were as expected, being a good starting point to adapt the work done here to create your own professional solution.
```

Reviewer 2: Please add a Next Step subsection inside conclusions as you did with the nicla vision streaming tutorial, refer to the tutorials that serves as a base for thsi tutorial and additional resources that you migh find useful as the whole PLC IDE docs software page with the tutorials
